working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1404,30 +1404,69 @@ To orchestrate this, type inference on patterns proceeds in three phases:
     a type *schema* and not a *type* because there may be holes where types
     aren't known yet.
 
+    We only calculate a pattern type schema for pattern variable declarations
+    and pattern assignments. In matching contexts (switch cases and if-case
+    statements), the pattern context type schema is not used, no downwards
+    inference is performed from the pattern to the matched value expression, and
+    no coercions or casts from `dynamic` are inserted in the matched value
```

Comment:
This is a breaking change from current behavior.  Probably not actually breaking in practice, but worth calling out, and definitely needs tests since otherwise the existing behavior will certainly carry over.

Reply:
Scratch this - I misread this as being scrutinee -> pattern, but you're talking about the other way around here.

Reply:
Lasse and I are discussing this choice [here](https://github.com/dart-lang/language/pull/2541#discussion_r1004222265). Do you think we should perform downwards inference on constant patterns? I'm fairly agnostic and can't really remember most of the arguments for or against.

Reply:
I don't know.  We currently do (for more than just conversions):

```dart
class Foo<T> {
  final T x;
  const Foo(T this.x);
}


test(Foo<num> f) {
  switch(f) {
    case Foo(3): print("equal"); break;
    default: print("not equal"); break;
  }
}
  
void main() {
  test(const Foo(3));
}
```

That said, I don't suspect it's actually that useful in practice.  

One thing that we might want to think about is what we want to happen if we add user defined coercions.  That is, if we, say, added implicit constructors, and had an implicit constructor which constructed a `Foo` from an `int` if an `int` was used where a `Foo` was expected, what would we want:

```dart
test(Foo f) {
  switch(f) {
    case 3: ...
  }
}
```

to do?

In general, I don't love the coercions here.  I think my preference would probably be to say:

1) No coercions in refutable patterns
2) Yes downwards inference on constants in refutable patterns

Reply:
Your preferences SGTM. Done.